### PR TITLE
Nicer buffer handling

### PR DESCRIPTION
**Pull request by original-brownbear**
Some optimisations that I found when reusing searchable snapshot code elsewhere:
* Add an efficient input stream -> byte buffer path that avoids allocations + copies for heap buffers, this is non-trivial in its effects IMO
  * Also at least avoid allocations and use existing thread-local buffer when doing input stream -> direct bb
  * move `readFully` to lower level streams class to enable this
* Use same thread local direct byte buffer for frozen and caching index input instead of constantly allocating new heap buffers and writing those to disk inefficiently

### CODE REVIEW

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/input/FrozenIndexInput.java [hunk 1]
```diff
@@ -255,28 +239,19 @@ private void writeCacheFile(
...
         );
         final long end = relativePos + length;
-        final byte[] copyBuffer = new byte[toIntBytes(Math.min(COPY_BUFFER_SIZE, length))];
         logger.trace(() -> format("writing range [%s-%s] to cache file [%s]", relativePos, end, cacheFile));
 
         long bytesCopied = 0L;
         long remaining = length;
-        final ByteBuffer buf = writeBuffer.get();
-        buf.clear();
+        final ByteBuffer buf = writeBuffer.get().clear();
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/input/FrozenIndexInput.java:246]: Just a nit but I wonder if we could make `writeBuffer` private and add a protected `getWriteBuffer()` in `MetadataCachingIndexInput` that ensures the `ByteBuffer` is properly cleared before usage? We could also check which threads are accessing it.

**Review by tlrx — APPROVED** [review]
LGTM. I haven't seen any issues and it looks better indeed. I left a minor comment.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/input/FrozenIndexInput.java (hunk 1 above)
Comment (original-brownbear) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/input/FrozenIndexInput.java:246]: I'm always a little hesitant hiding the fact that something is thread-local behind another layer to be honest. I agree it looks nicer here, but it makes it much easier to take a wrong step for someone new to this code here.

**Comment by original-brownbear** [comment]
Thanks Tanguy!

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/input/FrozenIndexInput.java (hunk 1 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/input/FrozenIndexInput.java:246]: Yeah, that's a valid argument too.

**Pull request merged**